### PR TITLE
Clip perf tune

### PR DESCRIPTION
**Pull request by x574chen**
# What does this PR do?

Improves Clip performance on Gaudi2. Below is the performance of batch size 512 removing checkpoint saving time, which is measured using Habana 1.10 release docker version on Gaudi2.

image

### CODE REVIEW

**Comment by x574chen** [comment]
Regarding the [mediapipe_dataloader argument](https://github.com/huggingface/optimum-habana/pull/256/files#diff-f02f4e167ba4acd9a11c47de24ce5252a80e5a8bb6321d21ad0efab6ac120099R170) in 8x command , the loss for 20 epoch compared with the pytorch dataloader is attached.



Since the [resize in habana media dataloader](https://github.com/huggingface/optimum-habana/pull/256/files#diff-de4ae9309337ddf655cdd28451c1128420dca18cc3a7e4120222c98b201552aaR143)  is not the same with [resizing in pytorch dataloader](https://github.com/huggingface/optimum-habana/blob/main/examples/contrastive-image-text/run_clip.py#LL197C13-L197C19), it is expected that their losses are slightly different. More details about their behavior, see [torchvision Resize](https://pytorch.org/vision/main/generated/torchvision.transforms.Resize.html)  and [habana Resize](https://docs.habana.ai/en/latest/Media_Pipeline/Decoder.html).

File: examples/contrastive-image-text/clip_media_pipe.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    [],
+    3,
+    read_image_text_from_dataset_params,
+    None,
+    read_image_text_from_dataset,
+    dtype.NDT,
+)
+op_class = fn.operator_add("ClipDataReader")
+op_class.__module__ = fn.__name__
+setattr(fn, "ClipDataReader", op_class)
```
Comment (regisss) [inline examples/contrastive-image-text/clip_media_pipe.py:103]: Is it planned to add an image-text reader to `habana_frameworks.mediapipe` in a future version of SynapseAI?

File: examples/contrastive-image-text/clip_media_pipe.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        )
+
+        ClipMediaPipe.instance_count += 1
+
+    def definegraph(self):
+        res_pp_filter = ftype.BICUBIC
+        self.input = fn.ClipDataReader(label_dtype=dtype.UINT32, dataset=self.dataset)
+        jpegs, input_ids, attention_masks = self.input()
+        def_output_image_size = [self.image_size, self.image_size]
+        self.decode = fn.ImageDecoder(
```
Comment (regisss) [inline examples/contrastive-image-text/clip_media_pipe.py:138]: Same comment as for `self.input`

File: examples/contrastive-image-text/clip_media_pipe.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+            resize=def_output_image_size,
+            resampling_mode=res_pp_filter,
+        )
+        images = self.decode(jpegs)
+        cmn_pos_offset = 0.5
+        normalize_mean = np.array(self.dataset.image_mean).astype(np.float32) * 255
+        normalize_std = 1 / (np.array(self.dataset.image_std).astype(np.float32) * 255)
+        norm_mean = fn.MediaConst(data=normalize_mean, shape=[1, 1, 3], dtype=dtype.FLOAT32)
+        norm_std = fn.MediaConst(data=normalize_std, shape=[1, 1, 3], dtype=dtype.FLOAT32)
+        cmn = fn.CropMirrorNorm(
```
Comment (regisss) [inline examples/contrastive-image-text/clip_media_pipe.py:151]: Same

File: examples/contrastive-image-text/run_clip.py [hunk 4]
```diff
@@ -57,6 +58,132 @@
...
+        if self.fallback_activated:
+            return super().__next__()
+
+        data = next(self.iterator)
+        return {
+            "pixel_values": data[0],
+            "input_ids": data[1],
+            "attention_mask": data[2],
+            "return_loss": True,
+        }
```
Comment (regisss) [inline examples/contrastive-image-text/run_clip.py:128]: I would move the definition of this class to `clip_media_pipe.py`

File: examples/contrastive-image-text/run_clip.py [hunk 5]
```diff
@@ -57,6 +58,132 @@
...
+
+
+class HabanaDataloaderTrainer(GaudiTrainer):
```
Comment (regisss) [inline examples/contrastive-image-text/run_clip.py:131]: Let's move the definition of this class to a new file called `habana_dataloader_trainer.py` (same as for the QA exemple [here](https://github.com/huggingface/optimum-habana/tree/main/examples/question-answering) actually)

File: examples/contrastive-image-text/run_clip.py [hunk 6]
```diff
@@ -57,6 +58,132 @@
...
+    def get_train_dataloader(self):
+        """
+        Returns the training [`~torch.utils.data.DataLoader`].
+
+        Will use no sampler if `train_dataset` does not implement `__len__`, a random sampler (adapted to distributed
+        training if necessary) otherwise.
+
+        Subclass and override this method if you want to inject some custom behavior.
```
Comment (regisss) [inline examples/contrastive-image-text/run_clip.py:139]: You can remove this and replace it by a one-line explanation of what is different from the regular trainer.

File: examples/contrastive-image-text/run_clip.py [hunk 7]
```diff
@@ -99,6 +226,7 @@ class ModelArguments:
     freeze_text_model: bool = field(
         default=False, metadata={"help": "Whether to freeze the text model parameters or not."}
     )
+    mediapipe_dataloader: bool = field(default=False, metadata={"help": "Turn on MediaPipe HW dataloading"})
```
Comment (regisss) [inline examples/contrastive-image-text/run_clip.py:229]: This argument is not related to the model, let's move it to `DataTrainingArguments` right below.

File: examples/contrastive-image-text/run_clip.py [hunk 8]
```diff
@@ -260,6 +388,7 @@ def main():
         revision=model_args.model_revision,
         use_auth_token=True if model_args.use_auth_token else None,
     )
+    gaudi_config.use_fused_adam = False
```
Comment (regisss) [inline examples/contrastive-image-text/run_clip.py:391]: Out of curiosity, why disabling fused ADAM? Is it not compatible with the mediapipe?

**Review by regisss — COMMENTED** [review]
Nice speedup with this PR @x574chen!!

I left a few comments to address.
Also, HPU graphs for training are going to be integrated into the lib by #200, so let's wait for it to be merged and reuse it here.

File: optimum/habana/transformers/trainer.py [hunk 9]
```diff
@@ -387,6 +387,14 @@ def _wrap_model(self, model, training=True, dataloader=None):
...
             return model
 
+        if self.args.use_hpu_graphs:
+            import habana_frameworks.torch as ht
+
+            model = ht.hpu.ModuleCacher()(
+                model=model, inplace=True, allow_unused_input=True, have_grad_accumulation=True
+            )
+            print("INFO: Using HPU Graph for training...")
+
```
[the remark this answers is not in the document]
Comment (x574chen) [reply optimum/habana/transformers/trainer.py]: change for hpu graph training is removed.

File: examples/contrastive-image-text/clip_media_pipe.py (hunk 1 above)
Comment (x574chen) [reply examples/contrastive-image-text/clip_media_pipe.py:121]: No. Media reader is closely related to each task (image/text keys).  See [changes in clip reader](https://github.com/huggingface/optimum-habana/pull/256/files#diff-de4ae9309337ddf655cdd28451c1128420dca18cc3a7e4120222c98b201552aaR93).

File: examples/contrastive-image-text/clip_media_pipe.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    dtype.NDT,
+)
+op_class = fn.operator_add("ClipDataReader")
+op_class.__module__ = fn.__name__
+setattr(fn, "ClipDataReader", op_class)
+
+
+class ClipMediaPipe(MediaPipe):
+    """
+    Class defining clip media pipe.
```
[the remark this answers is not in the document]
Comment (x574chen) [reply examples/contrastive-image-text/clip_media_pipe.py]: fixed.

File: examples/contrastive-image-text/clip_media_pipe.py [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+        super(ClipMediaPipe, self).__init__(
+            device=self.device, batch_size=batch_size, prefetch_depth=queue_depth, pipe_name=pipe_name
+        )
+
+        ClipMediaPipe.instance_count += 1
+
+    def definegraph(self):
+        res_pp_filter = ftype.BICUBIC
+        self.input = fn.ClipDataReader(label_dtype=dtype.UINT32, dataset=self.dataset)
```
[the remark this answers is not in the document]
Comment (x574chen) [reply examples/contrastive-image-text/clip_media_pipe.py:152]: fixed.

File: examples/contrastive-image-text/clip_media_pipe.py (hunk 2 above)
Comment (x574chen) [reply examples/contrastive-image-text/clip_media_pipe.py:155]: fixed.

File: examples/contrastive-image-text/run_clip.py (hunk 4 above)
Comment (x574chen) [reply examples/contrastive-image-text/run_clip.py]: It cannot be moved to the same file with ClipMediaPipe. It will break the fallback mechanism if habana_frameworks.mediapipe is not installed. I move this class to a new file `clip_mediapipe_dataloader.py`.

File: examples/contrastive-image-text/run_clip.py (hunk 5 above)
Comment (x574chen) [reply examples/contrastive-image-text/run_clip.py]: fixed.

File: examples/contrastive-image-text/run_clip.py (hunk 6 above)
Comment (x574chen) [reply examples/contrastive-image-text/run_clip.py]: fixed.

File: examples/contrastive-image-text/run_clip.py (hunk 7 above)
Comment (x574chen) [reply examples/contrastive-image-text/run_clip.py]: fixed.

File: examples/contrastive-image-text/run_clip.py (hunk 8 above)
Comment (x574chen) [reply examples/contrastive-image-text/run_clip.py]: With mediapipe dataloader, using fused adam optimizer will OOM in case:
- Gaudi 2
- Batch size = 512

Comment (x574chen) [reply examples/contrastive-image-text/run_clip.py]: I add a condition of using mediapipe to disable fused adam

File: examples/contrastive-image-text/clip_media_pipe.py (hunk 3 above)
Comment (x574chen) [reply examples/contrastive-image-text/clip_media_pipe.py]: fixed.

File: examples/contrastive-image-text/run_clip.py [hunk 12]
```diff
@@ -167,6 +168,7 @@ class DataTrainingArguments:
         default=None,
         metadata={"help": "The number of processes to use for the preprocessing."},
     )
+    mediapipe_dataloader: bool = field(default=False, metadata={"help": "Turn on MediaPipe HW dataloading"})
```
Comment (regisss) [inline examples/contrastive-image-text/run_clip.py:171]:
```suggestion
    mediapipe_dataloader: bool = field(default=False, metadata={"help": "Turn on MediaPipe hardware-based accelerated data loading."})
```

**Review by regisss — APPROVED** [review]
Looks good to me!

I left a few very minor nits. I'll let you know when #200 is merged so that we can use it in this PR.

File: examples/contrastive-image-text/README.md [hunk 13]
```diff
@@ -108,7 +108,8 @@ python run_clip.py \
     --use_lazy_mode \
     --use_hpu_graphs \
     --gaudi_config_name Habana/clip \
-    --throughput_warmup_steps 3
+    --throughput_warmup_steps 3 \
+    --dataloader_num_workers 16
```
Comment (regisss) [inline examples/contrastive-image-text/README.md:112]: We don't use the mediapipe dataloader in 1x?

File: examples/contrastive-image-text/README.md [hunk 14]
```diff
@@ -136,7 +137,9 @@ python ../gaudi_spawn.py --world_size 8 --use_mpi run_clip.py \
     --use_lazy_mode \
     --use_hpu_graphs \
     --gaudi_config_name Habana/clip \
-    --throughput_warmup_steps 3
+    --throughput_warmup_steps 3 \
+    --dataloader_num_workers 16 \
+    --mediapipe_dataloader
```
Comment (regisss) [inline examples/contrastive-image-text/README.md:142]:
```suggestion
    --mediapipe_dataloader \
    --use_hpu_graphs_for_training \
    --distribution_strategy fast_ddp
```

**Review by regisss — COMMENTED** [review]
@x574chen Now that #200 got merged, we can add the training argument for using HPU graphs for training.

**Comment by regisss** [comment]
@x574chen The code quality check failed, please run the following from the root of the repo:
```bash
pip install black ruff
make style
```

File: examples/contrastive-image-text/README.md (hunk 13 above)
Comment (x574chen) [reply examples/contrastive-image-text/README.md:112]: For 1x, simply increase Pytorch dataloader workers can achieve the best performance on G2. Using media dataloader are no longer improving performance.

**Comment by regisss** [comment]
@x574chen Trying to work on another example (different dataset) using mediapipe, I got the following error:
```
recipeHandle is a null ptr synGraphCompile has failed
```
when instantiating `HPUGenericPytorchIterator(mediapipe=pipeline)`.
Do you have any idea what is going wrong here?

**Comment by x574chen** [comment]
@regisss  Are you working on Gaudi1? There is no Media IP in G1.

**Comment by regisss** [comment]
> @regisss Are you working on Gaudi1? There is no Media IP in G1.

Good call! I just tried on Gaudi 2 and it works :)
I missed that bit in Synapse documentation that it only works on Gaudi2 :sweat_smile: 

I'll make it clearer in the README of the CLIP example too.
